### ISSUE TITLE
build of 0.5.1 fails

### ISSUE THREAD

**Issue by rofl0r** [gbdev/rgbds#922]
```
src/asm/parser.y:531.1-11: invalid directive: ‘%precedence’
src/asm/parser.y:671.19-24: invalid directive: ‘%empty’
src/asm/parser.y:776.19-24: invalid directive: ‘%empty’
src/asm/parser.y:807.19-24: invalid directive: ‘%empty’
src/asm/parser.y:870.19-24: invalid directive: ‘%empty’  
src/asm/parser.y:951.19-24: invalid directive: ‘%empty’
src/asm/parser.y:1069.19-24: invalid directive: ‘%empty’
src/asm/parser.y:1606.19-24: invalid directive: ‘%empty’
src/asm/parser.y:1630.19-24: invalid directive: ‘%empty’
src/asm/parser.y:1645.19-24: invalid directive: ‘%empty’
src/asm/parser.y:1656.19-24: invalid directive: ‘%empty’
```

```
bison --version
bison (GNU Bison) 2.6.2
Written by Robert Corbett and Richard Stallman.
```

my bison version compiles everything else (an entire linux distro) fine - rgbds is the only program using these bleeding egde features.

**Comment by ISSOtm** [comment]
RGBDS 0.5.1+ requires Bison ≥ 3.0.0, which is from 2013, which we consider reasonable.
2.6.2 is, AIUI, the default on macOS (being the last GPLv2 version); we recommend using the version from Homebrew.

Is that fine with you?

**Comment by rofl0r** [comment]
no, that's why i opened pr #923 - there's absolutely no reason for these changes, and you slipped it in like a trojan horse under an innocent title - "enable bison warnings"

**Comment by ISSOtm** [comment]
The reason for these changes is to simplify the code (`%precedence` and then-redundant directive removal), and make modifications more reliable (`/* empty */` does not guard against a forgotten `|`, but `%empty` does). There *is* reason for these changes.

It was not "slipped in", those changes were in the spirit of the commit to improve reliability. These specific changes are not enabling warnings per se (as `-Wall` would do), but they are enabling better error reporting as described above, so I consider the topicality correct. (And even if it wasn't the case, mistakes happen occasionally.)

I was asking if "this is fine with you" to learn why using Bison ≥ 3.0 may not feasible or practical for you, and hopefully reach a satisfactory compromise between our goals and yours.

**Comment by daid** [comment]
Insert xkcd workflow.

**Comment by rofl0r** [comment]
> I was asking if "this is fine with you" to learn why using Bison ≥ 3.0 may not feasible or practical for you, and hopefully reach a satisfactory compromise between our goals and yours.

flex and bison are quite buggy codebases and it's difficult to find versions that work reliable across all packages in my distro, without going for day-long patch orgies. 2.6.2 works for all other packages so i'd like to avoid bumping it just for the sake of a single package.

**Comment by ISSOtm** [comment]
Alright. I have to run a couple errands, so I'll handle this in a few hours.

I see your use case, and I understand that you may want not to patch RGBDS (though I know a couple of downstream packagers, such as the BSDs, do it for us for other minor reasons).

I don't want to hard-revert those changes due to the benefits they provide, but maybe we can look into a `sed` invocation, if Bison < 3.0 is detected, that would replace `%empty` with the comments, etc. Can that work out?

**Comment by rofl0r** [comment]
> if Bison < 3.0 is detected, that would replace %empty with the comments, etc. Can that work out?

that could work for %empty, though i guess not for %precedence - as your commit removed the %left operator from everywhere else.

what could work is if parse.y is generated from a template (parse.y.in) using e.g. a preprocessor pass or similar on it.

**Comment by Rangi42** [comment]
I'm personally against making the parser file and the build process more complicated for the sake of supporting an eight-year-old `bison` version. macOS users should be used to upgrading various outdated built-in programs (IIRC they stick with some very old Linux utilities for license reasons), and Linux/BSD/etc users are capable of either patching parser.y to avoid bison 3 features, or patching the Makefile to use a local copy of bison 3. (I'm used to doing that myself with different rgbds versions for some old assembly projects.)

However, if we do end up adding a `sed` patching step, this could work:

- Update PR #923 to replace the "`%left`"s that it adds with "`%bison2_left`"
- In the Makefile, if the bison version is 3, use `sed` to remove all lines starting with "`%bison2_left`"
- If the bison version is 2, replace all "`%bison2_left`" with "`%left`", remove all "`%empty`", and remove all lines starting with "`%precedence`"

**Comment by aaaaaa123456789** [comment]
I think I've spoken loudly enough against Rust in every venue I've been afforded to not be mistaken for a Rust supporter, but I'll go ahead and say this: rewriting RGBDS in C++, Rust, or any other similar language is not inherently harmful.

I'm a user, not a maintainer. I've contributed very little code to RGBDS. And from a user's perspective, what matters is that the software (1) works, and works well, (2) is easy to build and update, and (3) can be installed on any platform I use. (1) and (3) are equally true of all those languages. (2) is fine for C++, but more problematic for Rust, because Rust loves being the odd one out and using a build process that is different from the rest of the universe (including _fetching packages from the Internet_ as part of the process!), but I assume skilled Rust devs could wrap this into a build process that is compatible with the world and doesn't require (or use) an active connection to the Internet.
The move means I won't be contributing in the future, sure — I don't know Rust and don't have any interest in learning it (I've tried enough times to realize I simply don't enjoy the language). But I'm not a contributor right now, so that has no impact. The same goes for literally any other non-contributing user out there. As a user, the language the tool is written in is irrelevant.

So I'd fully support moving to a language the maintainers are comfortable writing in, whatever that may be, as long as the conditions I mentioned above are preserved. In the case of Rust, that probably means writing a simple makefile that wraps `rustc` (or `cargo` if you must) and doesn't trigger online package fetches unless you tell it to.

**Comment by AntonioND** [comment]
I think that a transition to C++ is be far more realistic (and so, more likely to actually happen) than a complete rewrite in Rust. It would also keep the build system simple, which is crucial, in my opinion.

I kinda hate the kind of things that some people do with C++, though.

**Comment by rofl0r** [comment]
> in practice C has been a cause of rgbds issues (use-after-free, buffer overflows, null-terminated strings, having to manually malloc/free, undefined behavior, etc) not a solution to them

i doubt it's been C causing those issues, but programmers using it incorrectly...

> transition to C++, for the sake of using some of its features (in particular to unblock #885, but I expect RAII idioms will also make for cleaner code)

i see you already refactored the code to use a struct string, which seems like a good way forward. what i glanced from the PR it seems almost complete and there's probably not a lot left to change - a rewrite using C++ would be a whole lot more work.
coincidentally, i've been writing a similar string library in the past: https://github.com/rofl0r/libulz/blob/master/include/stringptr.h - maybe you can draw some inspiration from it.

btw, here's what linus torvalds thinks of C++: http://harmful.cat-v.org/software/c++/linus

**Comment by ISSOtm** [comment]
> i see you already refactored the code to use a struct string, which seems like a good way forward. what i glanced from the PR it seems almost complete and there's probably not a lot left to change - [...]

The problem is that what there *is* to change is making it work with the entirety of `hashmap.c`, which mind you, is already used elsewhere in the codebase (RGBLINK, to be precise).

**Comment by AntonioND** [comment]
> btw, here's what linus torvalds thinks of C++: http://harmful.cat-v.org/software/c++/linus

C++ is a bad language for a kernel for the simple fact that the language does memory management without your control. For example, I was once writing a firmware for a microcontroller, and someone decided to start adding C++ stuff. It took myself and another guy a couple of hours to convince him that his ``std:pair`` statically allocated array was pulling ``malloc`` and company into the firmware. So yeah, I'm not a fan of C++ fans.

However, If you just use C with classes and maybe with operator overloading and things like that... the result tends to be okay. Just avoid the C++ standard library... and almost every other library.

**Comment by rofl0r** [comment]
> Also the first file I checked in strlib, containsChar.c, could have just been return strchr(str, what) != NULL. (Which doesn't say much good about C's built-in "string" (aka pointer to some bytes which we hope end in a '\0') handling functions either.)

you're looking in the wrong place, the string+size struct stuff is provided by src/stringptr - and the purpose of strlib is to pull in as little of libc as possible, to keep static binaries small (iirc the only libc funcs called are memcpy and malloc* - and the latter only if a *_new func is used)

> However, If you just use C with classes and maybe with operator overloading and things like that... the result tends to be okay. Just avoid the C++ standard library... and almost every other library. I would use std::string and little else.

that's true, but once C++ is being used, you're automatically attracting C++ "wizards", which will start opening PRs using their favorite feature subset... just as linus points out in that article.

**Comment by ISSOtm** [comment]
Sure, but we can simply review out unnecessarily complex code from contributions. (Whether you trust us not to write such code ourselves is left as an exercise to the reader.)

As for the possible performance impact of such a decision, we do not have any performance regression tests yet because we haven't figured out how to do those, but please feel free to chime in to #653 if you have any ideas.
